쥬스 메이커 [STEP 1] sookoong, bard

Jeon-Minsu opened the pull request:
안녕하세요 Jake! @jryoun1 
이번 쥬스 메이커 프로젝트를 담당하게 된 수꿍과 바드입니다!!
STEP 1 기능 구현 Pull Request 드립니다!

## 고민한점
1) 과일과 재고 저장 타입을 `dictionary`로 할지 `tuple`로 할지 고민하였습니다. 이에 key, value를 통해 호출이 더 용이한 dictionary 타입이 이번 프로젝트에 적절하다 판단하여 이를 사용했습니다.

2) 과일재고의 타입을 `UInt`로 설정해도 되는지 고민이 되었습니다. 이에 두가지 근거를 가지고 `Int`타입 설정후 에러 핸들링을 통해 처리하였습니다. 첫째는 UInt에 대한 사용 경험이 많지 않아 기피되었습니다. 두번째로, 그 수가 확실히 양의 정수라는 확신이 없을 때에는 정수값에 대하여 Int값을 기본으로 설정하란 내용을 들은 것 같아 위와 같이 행동해보았습니다.

3) 과일의 재고가 부족하여 과일쥬스를 제조할수 없을 때 코드 구현 이전에는 조건문을 설정하고 반복문을 처리할까 생각했습니다. 하지만 이후 코드 구현 과정에 있어서 에러 처리의 필요성을 인식하였고, 또한 반복하여 작동할 필요성을 느끼지 못하여 guard문을 이용한 에러처리로 구현하였습니다.

4) 제조할 수 있는지 없는지를 판단하는 함수를 생성해야되는지 의문이 생겼습니다. 다른 과일 입력을 받을 것 같지는 않으나 좀 더 확장성 있는 코드를 목표시하여 정해진 과일 외 다른 과일로 쥬스를 제조할 때 오류로 핸들링할 수 있도록 기능을 구현하였습니다.

## 의문점
1) 이번 프로젝트 지침과 관련하여 `juiceMaker`는 `FruitStore`를 소유하고 있다는 말은 상속을 받는 것인지 아니면 `juiceMaker`안에 인스턴스로 구현하는 것을 의미하는지 확신이 들지 않았습니다. 

2) 이번 고민은 고민한점의 4번 고민과 연관되어 있습니다. 쥬스를 만들고자 할 때 정해진 쥬스만을  제작할 수도 있지만, 다른 과일로도 제작된 쥬스를 요청할 수 있기 때문에 파라미터로 다양한 입력을 받을 수 있어야 된다고 생각하였습니다. 그러다보니 처음에는 파라미터의 자료형을 Juice로 설정하여 열거형 Juice 내의 값을 간단히 입력받을 수 있었습니다. 
  하지만, 열거형으로 사용하다보니 case 이외의 값을 입력받았을 때는 빌드 자체가 구현되지 않았습니다. 이를 해결하기 위하여, 열거형 타입에 `extenstion`을 사용해 스트링값을 할당하니 정해진 경우 이외의 값을 입력받아도 오류처리를 통해 해결할 수 있었습니다. 여기서 과연 전자의 기능만을 구현하는 것으로 충분할 지, 아니면 후자의 기능까지 구현을 하는 것이 좋은지 의문이 듭니다.

3) 현재 `makeJuice`함수 내에서 현재 너무 많은 기능들이 하나의 함수에 구현되고 있는데 이 부분을 어떻게 나누는 게 적절할지 고민중입니다.

Jeon-Minsu committed:
feat: Fruit 열거형 및 재고 관련 에러 열거형 추가

bar-d committed:
feat: 과일과 재고를 저장할 딕셔너리 생성

Jeon-Minsu committed:
feat: 입력한 재고의 수로 변경하는 함수 생성

bar-d committed:
feat: 재고를 더하는 함수 생성

Jeon-Minsu committed:
feat: 입력한 수만큼 재고를 소모하는 함수 생성

bar-d committed:
feat: 주스 파일 및 열거형 생성

Jeon-Minsu committed:
refactor: 매직 넘버 및 매직 스트링 제거

bar-d committed:
refactor: 오류처리를 위한 코드 리팩토링

Jeon-Minsu committed:
feat: Fruit 열거형에 대한 extension 추가

bar-d committed:
feat: 주스를 만드는 함수 추가

Jeon-Minsu committed:
feat: stockError 열거형에 대한 extension 추가

bar-d committed:
refactor: StockError 열거형 파일 생성 및 주스 만드는 함수 에러처리 위해 리팩토링

Jeon-Minsu committed:
feat: 주스 메이커 내 makeJuice 함수의 에러 메시지 출력 구현

bar-d committed:
refactor: addStock함수 내 옵셔널에 대한 오류처리 구현

bar-d committed:
feat: 주스를 만드는데 사용되는 과일 갯수를 저장한 열거형 파일 생성

Jeon-Minsu committed:
feat: 초기 과일 재고값을 설정해주는 열거형 파일 생성

bar-d committed:
refactor: addStock 함수 내 옵셔널 처리 기능 중복되는 부분 삭제

Jeon-Minsu committed:
feat: 주스 열거형에 대한 extension 추가

bar-d committed:
docs: README.md 파일 생성 및 작성

Jeon-Minsu committed:
feat: 프로퍼티 은닉화

bar-d committed:
docs: README.md 파일 수정

Jeon-Minsu committed:
style: Dictionary 컨벤션 수정

bar-d committed:
docs: README.md 파일 수정

Jeon-Minsu committed:
refactor: rethrows를 통하여 makeJuice 함수 내 기능 분리

bar-d committed:
refactor: checkStock 함수 은닉화

Jeon-Minsu committed:
refactor: makeJuice 함수 내 에러처리 구문을 외부 함수를 통해 구현

bar-d committed:
style: StockError 열거형 내 invalidsection case 네이밍 변경 및 maekJuice함수 내 전달인자 레이블 생성

jryoun1 commented on JuiceMaker/JuiceMaker/Controller/AppDelegate.swift:
```diff
@@ -1,6 +1,6 @@
 //
 //  JuiceMaker - AppDelegate.swift
-//  Created by yagom. 
```
불필요한 코드를 제대로 수정해주셨네요. 좋습니다!!

jryoun1 commented on JuiceMaker/JuiceMaker/Model/Fruit.swift:
```diff
@@ -0,0 +1,31 @@
+    case mango
+}
+
+extension Fruit {
```
extension을 사용해서 분리를 하셨는데, extension을 사용하는 기준이 있을까요?

Jeon-Minsu replied:
`Fruit` 열거형의 case들에 String 값을 할당해주는 방법을 고민하였습니다. `Fruit` 열거형 내에서 바로 원시값으로 할당해줄 수도 있지만, name이라는 변수에 할당을 하는 동시에 열거형 구조 내에서 선언하는 것보다는 `extension`을 사용하는 방법이 더욱 가독성에 뛰어나다고 판단하여 이에 사용하였습니다.
물론 이후에 다시 언급이 되겠지만, `Fruit` 열거형 내에서만 과일의 종류를 사용하면 되기 때문에 extension을 삭제할 것입니다.

jryoun1 replied:
사용하는데 기준이 있고 그걸 공통적으로 지키는게 좋네요! 👍🏻

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitConsumption.swift:
```diff
@@ -0,0 +1,19 @@
+    static let mangoInMangoJuice = 3
+    static let mangoInMangoKiwiJuice = 2
+    static let kiwiInMangoKiwiJuice = 1
+    
```
여기 줄바꿈은 의도하신걸까요?

bar-d replied:
의도치 않은 공백이 있었네요🥲  
수정하도록하겠습니다

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitConsumption.swift:
```diff
@@ -0,0 +1,19 @@
+//  Created by 수꿍, 바드 on 2022/04/26.
+//
+
+enum FruitConsumption {
```
아래 상수들은 과일주스를 만드는데 소비되는 양인것 같은데
`FruiteConsumption` 은 과일 소비라고 해석되어 명확하지 않은 것 같다고 느껴지네요.
더 명확한 네이밍은 없을까요? 🤔

Jeon-Minsu replied:
`FruitConsumption`을 소비의 의미보다는 소모량의 의미로 사용했는데 위와 같이 의미가 명확하지 않을 수 있네요.
`CostOfJuiceProduction`으로 변경하여 쥬스 생산에 드는 비용임을 밝혀보았습니다.
여기서 제작비용의 비용이란 표현이 돈 뿐만 아니라 제작에 들어가는 재료 또한 이에 해당하는 의미로 사용해보았는데
이 표현 또한 명확하지 않다고 판단되시면 말씀해주세요!!

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,36 @@
 
-import Foundation
-
 // 과일 저장소 타입
```
이 주석이 필요해서 남겨두신 걸까요? 👀

bar-d replied:
수정해두었습니다
`Foundation`을 커밋 중간에 지워서 남아있던것 같습니다
FruitStore 파일 내에서는 `Foundation`이 필요가 없기에 지웠습니다🙃

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,36 @@
-
 // 과일 저장소 타입
 class FruitStore {
+    private var fruitDictionary: Dictionary<String, Int> = [Fruit.strawberry.name: InitialFruitSetting.stock, Fruit.banana.name: InitialFruitSetting.stock, Fruit.pineapple.name: InitialFruitSetting.stock, Fruit.kiwi.name: InitialFruitSetting.stock, Fruit.mango.name: InitialFruitSetting.stock]
```
private을 사용해 외부에서 사용하지 못하게 하셨네요! 좋습니다. 👍🏻
`fruiteDictionary` 는 과일 사전으로 해석되는데 과일 쥬스 레시피라고 부르는게 명확해보이네요.
dictionary 타입이라고 반드시 dictionary를 네이밍에 붙일 필요는 없다고 생각합니다.
필요에 따라서 붙일수도 있지만 의미를 한 번 생각해봐주세요.

그리고 가독성을 위해서 dictionary를 초기화 할 때 아래와 같이 작성하는건 어떤가요? 
```swift 
private var fruitDictionary: Dictionary<String, Int> = [
        Fruit.strawberry.name: InitialFruitSetting.stock,
        Fruit.banana.name: InitialFruitSetting.stock,
        Fruit.pineapple.name: InitialFruitSetting.stock,
        Fruit.kiwi.name: InitialFruitSetting.stock,
        Fruit.mango.name: InitialFruitSetting.stock
]
```
💡 코드의 정렬이 필요할 때는 원하시는 라인이나 블록을 드래그 하고 `ctrl + I` 를 누르면 자동 정렬을 해줍니다!

bar-d replied:
`fruitRecipe`로 네이밍 변경해주었습니다  
제이크의 말대로 작성해보니 더 가독성이 좋은것 같네요👍

bar-d replied:
제이크 다시 생각해보니 과일을 한군데 모아서 창고에 모아둔다는 느낌을 생각해서 제이크가 말씀해주신 `과일 쥬스 레시피`보다는 `과일창고`라는 말이 맞는것 같아서 `fruitWarehouse`로 다시 네이밍해주었습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,36 @@
 class FruitStore {
+    private var fruitDictionary: Dictionary<String, Int> = [Fruit.strawberry.name: InitialFruitSetting.stock, Fruit.banana.name: InitialFruitSetting.stock, Fruit.pineapple.name: InitialFruitSetting.stock, Fruit.kiwi.name: InitialFruitSetting.stock, Fruit.mango.name: InitialFruitSetting.stock]
+    
+    func changeStock(fruit: String, amount: Int) {
```
매개변수로 fruite string을 그대로 입력 받게되면, 스펠링 에러와 같은 휴먼에러가 발생할 수 있지 않을까요? 
```swift
// 작성가능한 case example
// case 1
change(fruit: "strawberry", amount: 3)

// case 2
let strawberry = "strawberry"
change(fruit: strawberry, amount: 3)
```

만약 오타로 인해서 fruit이 starwberry로 들어왔다면 아래의 코드는 의도한대로 실행될 수 있을까요? 
```swift
fruitDictionary[fruit] = amount
```

위에서 enum을 사용해서 fruit의 종류를 선언하셨는데 이를 사용해보는건 어떨까요?

Jeon-Minsu replied:
기존에  `Fruit` 과 `Juice` 내에서 `String` 타입의 값을 할당하는 `extension`을 지워서 기존의 `String` 타입으로 설정되어 있던 부분들을 모두 `Fruit` 과 `Juice` 열거형으로 변경하였습니다. 이로 인하여 열거형에 설정되어 있는 값만 선택이 이루어지도록 수정하였습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,36 @@
+    
+    func addStock(fruit: String, amount: Int) throws {
+        guard let fruitStock = fruitDictionary[fruit] else {
+            throw StockError.missingProduct
```
에러처리도 구현해주셨군요! 멋지네요

질문이 있습니다🙋 . 
만약 위에서 예시처럼 fruit이 오타로 들어오게되면 에러처리가 진행될 것 같은데 이것도 missingProduct 인지 궁금합니다!

Jeon-Minsu replied:
`String` 타입으로 과일을 입력받던 부분을 열거형인 `Fruit`으로 변경하여 오타의 가능성을 줄여보았습니다!
일단 `guard let`을 통한 에러 처리는 혹시 몰라 남겨보았는데 이를 삭제하는 게 더 좋을지 여쭈어 봅니다!

jryoun1 replied:
바드와 수꿍은 위 guard문은 어떤 역할을 해주고 있으며, 해당 역할이 필요한지 안 필요한지에 대해서 어떻게 생각하시나요?

Jeon-Minsu replied:
위 `guard` 문은 현재 입력된 과일이 `fruitWarehouse` 에 선언한 과일에 포함이 되는지 확인하는 역할을 수행합니다. 위의 역할은 필요하다고는 생각합니다. 기존과 같이 `String`으로 파라미터를 받는 것이 아닌, 열거형으로 파라미터를 입력받기 때문에 위의 오류를 최대한 피할 수 있지만 혹시라도 철자를 잘못 입력할 수도 있기 때문에 옵셔널 값이 반환됩니다. 따라서 이를 옵셔널 바인딩 과정을 수행하고 에러로 처리할 수 있도록 구현해보았습니다.

jryoun1 replied:
수꿍과 바드가 필요하다고 생각되면 삭제하지 않아도 될 것 같아요! 👍🏻

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,36 @@
+            throw StockError.missingProduct
+        }
+        let changedStock = fruitStock + amount
+        
```
줄바꿈은 의도하신 걸까요?

bar-d replied:
가드에도 상수가 들어가고 상수끼리 붙이고 그 다음 줄은 띄우기로 컨벤션을 했습니다    
붙이는게 컨벤션일까요?🤔

jryoun1 replied:
아뇨! 그렇지는 않습니다. 컨벤션이라는 것은 팀에서 정하기 나름입니다.
단지 팀에서 정한다고해도 일반적으로 사용하는 컨벤션을 기준으로 정해지기 때문에 Swift style guide를 많이 참고하시는 것이 Swift의 일반적인 컨벤션에 익숙해질 수 있는 방법 중 하나일 것 같네요 🙂

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,36 @@
+        fruitDictionary[fruit] = amount
+    }
+    
+    func addStock(fruit: String, amount: Int) throws {
```
amount가 Int 타입이라면 -개 를 더해줄수도 있을까요? 
그렇다면 그건 addStock일까요 reduceStock일까요? 🤔

Jeon-Minsu replied:
`amount`에 음수의 값이 입력이 된다면 위와 같이 `addStock`과 `reduceStock` 함수의 구분이 모호해진다고 생각합니다.
따라서, `amount`가 0 이상이라는 조건을 다음과 같이 사용하여 위의 문제를 해결해보았습니다. 
```swift
guard amount >= 0 else {
    throw JuiceMakerError.invalidAmount
}
```

jryoun1 commented on JuiceMaker/JuiceMaker/Model/InitialFruitSetting.swift:
```diff
@@ -0,0 +1,10 @@
+//  Created by 수꿍, 바드 on 2022/04/26.
+//
+
+enum InitialFruitSetting {
```
과일의 초기 재고는 왜 enum으로 따로 분리가 되었는지 궁금합니다!
과일의 초기 재고는 누가 관리하는 변수일까요?

bar-d replied:
네임스페이스를 활용하기 위해 `enum`타입으로 만들었습니다  -> ~하드코딩을 피하기 위해서...!~
과일의 초기재고는 `FruitStore`가 관리하니 `FruitStore`파일 내에 넣어주시라는 말씀일까요?

jryoun1 replied:
앗 그런 것은 아니고 그냥 이유가 궁금해서 여쭤봤습니다. 🙏🏻
FruitStore 파일 보다는 FruitStore 클래스의 프로퍼티라고 하는 것이 오해의 소지가 없어보입니다.

jryoun1 replied:
이 열거형 타입의 네이밍도 초기 과일의 무엇을 셋팅하는 걸까요? 
`DefaultFruitAmount`, `InitialFruitStock` 등 과 같은 네이밍이 좀 더 명확하지 않을까요?

jryoun1 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,52 @@
 struct JuiceMaker {
+    var JuiceMakersStore = FruitStore()
+    
+    func makeJuice(of juice: String) {
```
`makeJuice(of: strawberryJuice)` 이런식으로 호출될 것 같은데 어색하다고 느껴지네요.
`make(strawberryJuice)` 나 `makeJuice(kind: strawberry)` 이런 네이밍은 어떤가요? 

네이밍은 항상 쉽지 않은 것 같네요 🥲

bar-d replied:
전자 방안으로 네이밍 변경했습니다🥲

jryoun1 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,52 @@
-
 // 쥬스 메이커 타입
 struct JuiceMaker {
+    var JuiceMakersStore = FruitStore()
```
변수명을 대문자로 사용하는 것은 Swift style guide에서는 권장하지 않습니다!
또한 변수명에 s는 배열에 자주 붙입니다.
```swift
let names = ["바드", "수꿍", "야곰"]
```
각 언어마다 네이밍에 대한 어느 정도의 규칙이 존재하고 이를 따르는 것이 협업에 있어서 많은 도움이 될 것 입니다. [Swift API Design Guidelines](https://swift.org/documentation/api-design-guidelines/) 문서를 꼼꼼히 읽어보시는 걸 추천드려요!

bar-d replied:
`juiceMaker의 store`라는 느낌을 주기위해 `juiceMakersStore`라는 변수를 주었습니다.  
대문자는 실수입니다😂  
`juiceMakerStore`로 변경해주었습니다.  
문서 다시한번 읽어보겠습니다📝
`storeOfJuiceMaker`도 나쁘지 않았을것 같네요

jryoun1 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,52 @@
-
 // 쥬스 메이커 타입
 struct JuiceMaker {
+    var JuiceMakersStore = FruitStore()
```
var 와 let을 사용하는 기준이 있을까요?

bar-d replied:
`var`는 값이 변경될 때, `let`은 값이 변경이 없을 때 사용한다고 생각합니다  
`juiceMakerStore`는 클래스 참조 타입에서 가져오는 인스턴스이기 때문에 `let`으로 변경해도 괜찮을것 같습니다!

jryoun1 commented on JuiceMaker/JuiceMaker/Model/StockError.swift:
```diff
@@ -0,0 +1,22 @@
+//  Created by 수꿍, 바드 on 2022/04/26.
+//
+
+enum StockError: Error {
```
에러 처리를 위한 에러 enum 타입도 구현하셨군요 👍🏻
[LocalizedError](https://developer.apple.com/documentation/foundation/localizederror) 에 대해서도 한 번 공부해보시면 좋을 것 같습니다.

bar-d replied:
공부해보겠습니다✏️

Jeon-Minsu replied:
`LocalizedError`를 사용하여 기존의 에러 처리 방식을 수정해보았습니다!

jryoun1 commented on JuiceMaker/JuiceMaker/Model/StockError.swift:
```diff
@@ -0,0 +1,22 @@
+//  Created by 수꿍, 바드 on 2022/04/26.
+//
+
+enum StockError: Error {
```
`StockError`은 재고에 관련된 에러만 다루는 걸까요?
만약 쥬스메이커의 시스템 에러가 발생한다면 이 에러는 재고 에러에 포함될 수 있을까요?

bar-d replied:
`FruitStoreError`으로 변경해두었습니다

jryoun1 commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -0,0 +1,37 @@
+//
+
+enum Juice {
+    case strawberryJuice
```
이 Juice 타입을 사용할 때 
`Juice.strawberryJuice.name` 이런식으로 사용이 될텐데 코드에 juice의 반복이 많아보이네요.

bar-d replied:
`Juice`타입을 `Product`로 네이밍을 변경하였습니다

jryoun1 replied:
아래 코멘트 드렸습니다!

jryoun1 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,52 @@
+        }
+    }
+    
+    private func checkStock(function: (String, Int) throws -> Void, juice: String) rethrows {
```
함수를 매개변수로 넘기셨네요 멋지네요. 👍🏻
이 함수가 고민에 의해서 생성된 함수인 것 같네요. 

개인적인 생각이니 참고만 해주세요! 
**개인적인 생각**으로는 정해진 과일이외에 다른 과일이 들어올 수 있기 때문에 확장성을 위해서 구현하신것은 충분히 고민해볼만 하지만 오버엔지니어링이라고 생각합니다.
그 이유는 기획에서 어떠한 과일 종류들이 사용 될 것이라고 나와있기 때문에 기획을 따르는 개발이 진행되는 것이 좋을 것 같다고 생각합니다. 기획은 언제 어떻게 바뀔지 모르기 때문에 확장성을 고려해 작성한 코드가 사용되지 않을수도 있습니다. 

물론 상황에 따라서 다르긴하고 확장성을 고려해서 코드를 작성할지 결정하는 것은 개발자 개인의 판단이니 수꿍과 바드의 의견에 따라서 결정하시면 될 것 같습니다!

Jeon-Minsu replied:
너무 확장성의 측면에 몰입되어 기존의 기획에 엇나간 것 같습니다. 저희도 쥬스를 호출할 때에 있어서 열거형의 case로 불러오는 편이 더욱 확실하고 편리하게 호출할 수 있다는 점에 공감합니다. 위의 조언을 받아들인 결과, `extension` 을 제거하여  `Product` 내에서의 쥬스만을 선택할 수 있도록 변경하였습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,52 @@
+        }
+    }
+    
+    private func checkStock(function: (String, Int) throws -> Void, juice: String) rethrows {
```
질문이 있습니다. 🙋
JuiceMaker가 stock을 체크하는 것이 나을까요 아니면 FruiteStore에서 stock을 체크하는것이 나을까요?
물론 현실과 코딩의 세계를 동일선상에 놓고 볼 수는 없지만 누가 어떤 역할을 가지는 것이 더 바람직한 지는 한 번쯤은 고민해볼만 할 것 같아요.

만약 이 함수의 의도가 **과일 주스를 만들 수 있는지 판단하는 함수**라면 네이밍이 명확하지 않을 것 같네요.
과일주스를 만들지 있는지 판단하는 함수라면 아래와 같은 로직들이 포함되어야하지 않을까요? 👀
- 과일의 종류가 있는지 확인
- 과일의 종류가 있다면 주문할 과일주스에 필요한 개수만큼 있는지 확인

Jeon-Minsu replied:
기존에 `JuiceMaker` 구조체 안에 내재한 `checkStock()` 함수와 `handle()` 함수를 `FruitStore` `Class` 내에 구현하여 `JuiceMaker` 내에서 `make()` 함수를 호출하면 `FruitStore`로 넘어가 재고를 확인하여 쥬스 제작 가능 여부를 판단하고 불가능할 시 에러를 처리하는 방식으로 수정하였습니다. 
이를 통해 `FruitStore` 내에서 과일의 종류가 있는지 확인은 물론 `Fruit` 열거형 내의 `case` 만을 다루지만 `checkMaterial()` 함수 내 `guard` 구문을 이용하여 과일의 종류가 있는지 확인을 하였고, 과일의 종류가 있다면 주문할 과일주스에 필요한 개수만큼 있는지 확인하는 기능은 `checkPossibilityOfMaking()` 함수를 통해 식별해보았습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,52 @@
+        }
+    }
     
+    private func handle(_ error: Error) {
```
에러 처리까지 구현이 되었군요 멋지네요!

bar-d replied:
감사합니다🙇‍♂️

jryoun1 replied:
모든 코멘트에 댓글을 달아야하는 부담을 안가지셔도 됩니다! 😂

jryoun1 reviewed (changes requested):
안녕하세요 수꿍 바드!
리뷰에 앞서서 아래 사항을 명심해주시면 감사하겠습니다. 🙇🏻‍♂️

리뷰어의 코드 리뷰는 정답이 아니며 개인적인 의견이니까 수꿍과 바드의 의견도 편하게 말씀해주세요 🙏🏻
작성한 코드에 대해서 근거가 명확하고 타당하다면 코드를 반드시 수정할 필요 없습니다. 
(댓글로 근거를 잘 표현해주세요!) 
또한 리뷰 내용이 잘 이해되지 않는 부분이 있다면 언제든지 말씀해주세요!

---- 

### 고민한 부분
> 1. 과일과 재고 저장 타입을 dictionary로 할지 tuple로 할지 고민하였습니다. 이에 key, value를 통해 호출이 더 용이한 dictionary 타입이 이번 프로젝트에 적절하다 판단하여 이를 사용했습니다.

tuple과 dictionary의 차이에 대해서도 한 번 알아보면 좋을 것 같습니다. 호출 외에도 값의 변경에 대해서도 알아보시면 좋을 것 같아요! 

> 2. 과일재고의 타입을 UInt로 설정해도 되는지 고민이 되었습니다. 이에 두가지 근거를 가지고 Int타입 설정후 에러 핸들링을 통해 처리하였습니다. 첫째는 UInt에 대한 사용 경험이 많지 않아 기피되었습니다. 두번째로, 그 수가 확실히 양의 정수라는 확신이 없을 때에는 정수값에 대하여 Int값을 기본으로 설정하란 내용을 들은 것 같아 위와 같이 행동해보았습니다.

사과가 -1개 있다는 의미는 어떤 의미인가요? 개인적인 생각으로는 과일 재고는 음의 정수가 될 수 없을 것 같습니다.

> 3. 과일의 재고가 부족하여 과일쥬스를 제조할수 없을 때 코드 구현 이전에는 조건문을 설정하고 반복문을 처리할까 생각했습니다. 하지만 이후 코드 구현 과정에 있어서 에러 처리의 필요성을 인식하였고, 또한 반복하여 작동할 필요성을 느끼지 못하여 guard문을 이용한 에러처리로 구현하였습니다.

반복으로 동작하지 않는 것과 guard문으로의 에러처리는 어떠한 관계가 있을까요? 
수꿍과 바드는 if와 guard를 어떠한 기준으로 사용하시나요?

> 4. 제조할 수 있는지 없는지를 판단하는 함수를 생성해야되는지 의문이 생겼습니다. 다른 과일 입력을 받을 것 같지는 않으나 좀 더 확장성 있는 코드를 목표시하여 정해진 과일 외 다른 과일로 쥬스를 제조할 때 오류로 핸들링할 수 있도록 기능을 구현하였습니다.

이 부분은 아래 리뷰에서 답변드렸습니다!

---

### 의문점 
> 1. 이번 프로젝트 지침과 관련하여 juiceMaker는 FruitStore를 소유하고 있다는 말은 상속을 받는 것인지 아니면 juiceMaker안에 인스턴스로 구현하는 것을 의미하는지 확신이 들지 않았습니다.

두 분은 어떠한 것이라고 생각하시는지와 그 이유가 궁금합니다! 

> 2. 이번 고민은 고민한점의 4번 고민과 연관되어 있습니다. 쥬스를 만들고자 할 때 정해진 쥬스만을 제작할 수도 있지만, 다른 과일로도 제작된 쥬스를 요청할 수 있기 때문에 파라미터로 다양한 입력을 받을 수 있어야 된다고 생각하였습니다. 그러다보니 처음에는 파라미터의 자료형을 Juice로 설정하여 열거형 Juice 내의 값을 간단히 입력받을 수 있었습니다.
하지만, 열거형으로 사용하다보니 case 이외의 값을 입력받았을 때는 빌드 자체가 구현되지 않았습니다. 이를 해결하기 위하여, 열거형 타입에 extenstion을 사용해 스트링값을 할당하니 정해진 경우 이외의 값을 입력받아도 오류처리를 통해 해결할 수 있었습니다. 여기서 과연 전자의 기능만을 구현하는 것으로 충분할 지, 아니면 후자의 기능까지 구현을 하는 것이 좋은지 의문이 듭니다.
 
이 부분도 아래 코드리뷰에서 답변드렸습니다.

> 3. 현재 makeJuice함수 내에서 현재 너무 많은 기능들이 하나의 함수에 구현되고 있는데 이 부분을 어떻게 나누는 게 적절할지 고민중입니다.

makeJuice 는 어떠한 역할을 하는 함수인지를 먼저 정의해보면 어떨까요? 
예를 들면 `makeFood()` 라는 함수가 있다면 이 함수는 음식을 만드는 함수일 것 같네요.
그렇다면 내부의 과정은 만드려고 하는 
1. 음식의 레시피를 체크
2. 필요한 음식 재료가 있는지 체크 
3. 필요한 음식 재료의 수량이 있는지 체크

이러한 과정들이 해당 함수에 포함이 될 수 있을 것 같아요. 
그렇다면 위의 과정들이 하나의 함수가 될 수 있고 `makeFood()`는 이러한 과정들을 캡슐화시키는 함수라고 생각할 수 있을 것 같아요. 
캡슐화에 대해서 더 자세히 공부해보시는 것을 추천드려요 🙏🏻

bar-d committed:
style: 네이밍 변경 및 줄바꿈 제거
- StockError에서 JuiceMakerError로 네이밍 변경
- fruitDictionary에서 fruitRecipe로 네이밍 변경
- makeJuice에서 make로 네이밍 변경
- Juice에서 Product로 네이밍 변경
- JuiceMakersStore에서 juiceMakerStore로 네이밍 변경

Jeon-Minsu committed:
refactor: 변수 네이밍 수정, 열거형 내 extension 제거, JuiceMakerError case 추가
- FruitConsumption에서 CostOfJuiceProduction으로 네이밍 수정
- Fruit과 Juice 열거형 내 extension 삭제
- extension 삭제로 인한 기존의 데이터 타입인 String을 Fruit과 Juice로 전환
- JuiceMakerError 내 invalidAmount case를 추가하여 amount 값이 0보다 작을 때 호출

bar-d committed:
refactor: 에러처리하는 열거형 LocalizedError로 익스텐션 리팩토링

Jeon-Minsu committed:
refactor: JuiceMaker 구조체 내 checkStock() 함수와 handle() 함수를 FruitStore Class 내부로 이동

bar-d committed:
refactor: 쥬스를 만들수 있는지 확인하는 함수(checkPossibilityOfMaking)를 생성하여 make 함수로 부터 분리

Jeon-Minsu committed:
refactor: 쥬스의 재료 과일이 있는지 확인하는 함수(checkMaterial)을 생성하여 make 함수 내 호출

bar-d committed:
style: 네이밍 변경 및 전달인자 레이블 생성
- fruitRecipe에서 fruitWarehouse로 변경
- checkStock(function: (Fruit))에서 goToWarehouse(of function: (Fruit))으로 변경

Jeon-Minsu commented:
안녕하세요 제이크!
이번 리뷰를 통해 좀 더 깊이 공부할 수 있는 내용들이 많았습니다! 감사합니다!
아래의 내용은 종합 코멘트에 관한 답변들입니다!

## 고민한 부분
>과일과 재고 저장 타입을 dictionary로 할지 tuple로 할지 고민하였습니다. 이에 key, value를 통해 호출이 더 용이한 dictionary 타입이 이번 프로젝트에 적절하다 판단하여 이를 사용했습니다.
>> tuple과 dictionary의 차이에 대해서도 한 번 알아보면 좋을 것 같습니다. 호출 외에도 값의 변경에 대해서도 알아보시면 좋을 것 같아요!


먼저 가장 큰 범위에서 둘의 차이는 `dictionary`는 `array`, `set`과 더불어 컬렉션 타입에 해당하지만, `tuple`은 컬렉션 타입이 아닙니다. 다음으로 세부적인 부분을 비교해보자면, 먼저 `dictionary`는 `key`와 `value` 2가지 종류의 값을 담을 수 있지만, `tuple`은 그보다 더 많이 담을 수 있고, 심지어 `tuple` 내에서는 `function` 타입도 포함할 수 있습니다. `tuple`은 주로 임시적으로 값을 종속시킬 때 사용하는데 임시가 아닌 존속의 의미로 사용하고자 하면 `class`나 `struct`가 적합합니다. 마지막으로 `tuple`은 순서가 있고, 생성된 이후 변경이 불가능하지만, `dictionary`는 순서가 없으며, 생성 이후에도 값의 변경이 가능합니다.


> 과일재고의 타입을 UInt로 설정해도 되는지 고민이 되었습니다. 이에 두가지 근거를 가지고 Int타입 설정후 에러 핸들링을 통해 처리하였습니다. 첫째는 UInt에 대한 사용 경험이 많지 않아 기피되었습니다. 두번째로, 그 수가 확실히 양의 정수라는 확신이 없을 때에는 정수값에 대하여 Int값을 기본으로 설정하란 내용을 들은 것 같아 위와 같이 행동해보았습니다.
>> 사과가 -1개 있다는 의미는 어떤 의미인가요? 개인적인 생각으로는 과일 재고는 음의 정수가 될 수 없을 것 같습니다.


과일 재고가 음의 정수를 가질 수 없다는 부분에 동의합니다. `UInt` 자료형을 사용하면 더욱 확실하게 변수의 데이터 타입을 구분할 수 있을 것 같습니다. 저희는 `Int` 자료형으로 재고의 데이터 타입을 설정하였지만 에러 핸들링을 통해 음의 정수로 값이 변경되지 못하게 처리하여서 충분하다고 생각합니다.


> 과일의 재고가 부족하여 과일쥬스를 제조할수 없을 때 코드 구현 이전에는 조건문을 설정하고 반복문을 처리할까 생각했습니다. 하지만 이후 코드 구현 과정에 있어서 에러 처리의 필요성을 인식하였고, 또한 반복하여 작동할 필요성을 느끼지 못하여 guard문을 이용한 에러처리로 구현하였습니다.
>> 반복으로 동작하지 않는 것과 guard문으로의 에러처리는 어떠한 관계가 있을까요?
수꿍과 바드는 if와 guard를 어떠한 기준으로 사용하시나요?


반복문을 처음에 언급한 이유는 과일 쥬스를 한번만 만들지는 않을 것 같다고 판단하여 반복을 시킨후 재고가 부족하거나 없는 과일 종류의 주스를 요구받을 때 멈추게 하는 방법을 생각한 것이었습니다. 하지만, 반복문을 사용하여 조건에 부합하지 않을때 종료시키기 보다는 하나의 명령을 입력받았을 때 에러가 있는지 확인하고 그에 맞게 처리하는 방법이 더욱 적절하다 판단하여 이로 수정한 것이었습니다.
`if`와 `guard`는 두 가지 기준으로 사용을 고민하는 것 같습니다. 첫번째는 `early exit`을 하고 싶을 때 `guard`를 사용하고 그렇지 아니한 경우에는 `if`를 사용합니다. 두번째로 전역변수로 사용하고자 할 때는 `guard`를 지역변수만으로 함수를 작성하고자 할 때는 `if`를 고려합니다. 추가로 가독성의 측면에서 `guard`가 더욱 뛰어나다 판단하여 `guard`를 애용하는 편이긴 합니다.



## 의문점
>이번 프로젝트 지침과 관련하여 juiceMaker는 FruitStore를 소유하고 있다는 말은 상속을 받는 것인지 아니면 juiceMaker안에 인스턴스로 구현하는 것을 의미하는지 확신이 들지 않았습니다.
>> 두 분은 어떠한 것이라고 생각하시는지와 그 이유가 궁금합니다!


저희는 후자의 방법으로 `JuiceMaker` 구조체 내에서 `FruitStore`를 인스턴스로 생성하였습니다. `FruitStore` 를 현실에 비유하자면, 전세계에 `JuiceMaker`를 가지는 모든 사람이 하나의 공통된 `FruitStore`를 이용하는 것은 아니라고 생각합니다. 우리나라를 예를 들면, 경기 지부, 강원 지부, 충북 지부와 같이 각각의 `FruitStore` 지점들이 있다고 판단하였습니다. 따라서 자신이 사용하는 `FruitStore`만을 이용하기 때문에 `JuiceMaker` 내에서 자신이 이용하는 `FruitStore`를 선언해야 한다고 생각했습니다.


> 현재 makeJuice함수 내에서 현재 너무 많은 기능들이 하나의 함수에 구현되고 있는데 이 부분을 어떻게 나누는 게 적절할지 고민중입니다.
>> makeJuice 는 어떠한 역할을 하는 함수인지를 먼저 정의해보면 어떨까요?
예를 들면 makeFood() 라는 함수가 있다면 이 함수는 음식을 만드는 함수일 것 같네요.
그렇다면 내부의 과정은 만드려고 하는
음식의 레시피를 체크
필요한 음식 재료가 있는지 체크
필요한 음식 재료의 수량이 있는지 체크
이러한 과정들이 해당 함수에 포함이 될 수 있을 것 같아요.
그렇다면 위의 과정들이 하나의 함수가 될 수 있고 makeFood()는 이러한 과정들을 캡슐화시키는 함수라고 생각할 수 있을 것 같아요.
캡슐화에 대해서 더 자세히 공부해보시는 것을 추천드려요 🙏🏻


위의 방법을 착안하여 `makeJuice()` 함수 내에 원하는 쥬스를 제작하기 위해 필요한 재료 과일이 포함되어 있는지 체크하는 `checkMaterial()` 함수와 필요한 재료 과일의 수량이 적절한지 판단하는 `checkPossibilityOfMaking()` 함수를 `try` 해보고, 에러 발생시 이를 처리하는 `handle()` 함수를 구현하여 단계별 쥬스 제작 방법을 표현해보았습니다.


이외에 언급하지 않은 코멘트들은 모두 세부 코멘트 내역에서 처리해보았습니다! 👍

bar-d committed:
docs: README.md 파일 수정

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,112 @@
+    ]
+    
+    func changeStock(fruit: Fruit, amount: Int) throws {
+        guard amount >= 0 else {
```
amount를 검증을 해주는 코드는 반복이 되고 있는 것 같네요! 
반복이 되는 코드는 따로 함수로 뺄 수도 있을 것 같아요

bar-d replied:
checkInavalidAmount()라는 함수로 빼주었습니다

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,112 @@
+        fruitWarehouse[fruit] = changedStock
+    }
+    
+    func check(fruit: Fruit) throws {
```
함수명이 명확해보이지 않습니다. 
무엇을 체크하는건지 함수명만 봐서는 알 수 없는 것 같아요.

Jeon-Minsu replied:
`check(fruit: Fruit)` 에서 `checkFruitKind(_ fruit: Fruit)`으로 변경하였습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/Product.swift:
```diff
@@ -0,0 +1,16 @@
+//  Created by 수꿍, 바드 on 2022/04/26.
+//
+
+enum Product {
```
Product는 상품인 것 같아서 기존에 Juice가 더 명확해보이긴 합니다.
case에 들어가는 juice를 제거해도 Juice.strawberry , Juice.kiwi, Juice.strawberryBanna 도 어떤 쥬스인지 충분히 알 수 있을 것 같아요!

jryoun1 commented on JuiceMaker/JuiceMaker/Model/Product.swift:
```diff
@@ -0,0 +1,16 @@
+//
+//  Juice.swift
```
놓치기 쉬운 부분이지만 여기도 Product로 수정이 되었어야할 것 같네요!  
나중에 파일을 찾을 때 헷갈리지 않습니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,27 @@
-
 // 쥬스 메이커 타입
 struct JuiceMaker {
+    let juiceMakerStore = FruitStore()
```
juiceMakerStore는 쥬스를 만드는 가게인 것 같고, 
fruitStore는 과일가게 인 것 같은데 네이밍이 일치하지 않아서 오해가 있을 수 있을 것 같네요.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,27 @@
+        }
+    }
+    
+    private func checkMaterial(_ juice: Product) throws {
```
해당 함수명만봐서는 뭘 하는 함수인지 알 수 없을 것 같아요 🥲

Jeon-Minsu replied:
checkFruitAvailable(for juice: Juice)로 수정해보았습니다. 혹시 더 나은 네이밍이 있는지 궁금합니다!

jryoun1 replied:
지금의 네이밍도 충분히 명확해 졌네요! 
`checkAvailableOfMaking(_ juice: Juice)` 라는 함수명도 괜찮아 보이네요!

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStoreError.swift:
```diff
@@ -0,0 +1,27 @@
+    case invalidAmount
+}
+
+extension FruitStoreError: LocalizedError {
```
LocalizedError를 적용하셨군요 멋집니다👍🏻
LocalizedError와 Error의 차이점은 무엇이고, LocalizedError는 어떤 프로토콜을 채택했을까요?

Jeon-Minsu replied:
`Error`는 `throw`될 수 있는 오류 값을 나타내는 프로토콜입니다.
`LocalizedError`는 `Error` 프로토콜을 상속하여, `extension`을 사용함으로써 `Error`에 대한 로컬 메시지를 출력합니다.

jryoun1 replied:
LocalizedError가 Error 프로토콜을 채택한다면, 
```swift
enum FruitStoreError: Error {
    case missingProduct
    case outOfStock
    case invalidAmount
}
```
위의 코드에서 Error 대신 LocalizedError가 대체되어도 될 것 같네요!

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStoreError.swift:
```diff
@@ -0,0 +1,27 @@
+    public var errorDescription: String? {
+        switch self {
+        case .missingProduct:
+            return NSLocalizedString("해당 상품이 없습니다", comment: "Description of missing Product")
```
NSLocalizedString과 String의 차이점은 무엇일까요? 👀

Jeon-Minsu replied:
`LocalizedError`에서는 `errorDescription` 변수를 생성하여 `NSLocalizedString` 값을 반환한 후, `localizedDescription`을 통해 출력 가능합니다. 
`NSLocalizedString`은 NSLocalizedString(key, comment)로 정의되어 `key`의 값이 문자열 변수, `nil` 값, 빈 칸이면 `export`하지 않기 때문에 `localizedDescription`를 사용하면 `optional` 값이 아닌 `String` 값으로 처리되어 유용합니다.

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStoreError.swift:
```diff
@@ -0,0 +1,27 @@
+}
+
+extension FruitStoreError: LocalizedError {
+    public var errorDescription: String? {
```
여기서 public은 어떤 역할을 할까요?

Jeon-Minsu replied:
모듈 외부에서도 접근 가능하게 하기 위하여 public을 사용했는데, 보통 LocalizedError를 사용한 예시들에 많이 사용되어 한 번 사용해봤습니다. 위의 이유 이외에 다른 이유가 있으면 알려주시기를 바랍니다.

jryoun1 replied:
접근제한자는 가능한 좁은 범위에서 사용하도록 설정하고, 
외부에서의 접근이 필요하다고 판단되면 그때 필요한 범위까지만 열어주는 형태로 구현하는 것이 좋습니다.
그 이유는 처음부터 넓은 범위 수정의 가능성을 열어둔다면, 의도치 않은 곳에서 수정으로 인한 사이드이펙트가 발생할 수 있기 때문입니다.

추후에 접근제한자의 종류 및 어느 경우에 주로 사용하는지 공부해보면 좋을 것 같네요!

jryoun1 commented on JuiceMaker/JuiceMaker/Model/FruitStoreError.swift:
```diff
@@ -0,0 +1,27 @@
+//
+//  StockError.swift
```
여기도 놓치지말고 바꿔주세요!

jryoun1 reviewed (commented):
수꿍 바드 모두 고생하셨습니다!

jryoun1 commented:
### 고민한 부분에 대한 답변
1. 두 개의 차이를 잘 파악하셨네요 좋습니다! 👍🏻
2. 넵 좋습니다. UInt 타입으로 한다면 예외처리를 매번 하지 않아도 된다는 점과 이 코드를 처음보는 사람도 과일의 값은 확실히 음수가 될 수 없다는 것을 타입만 보고도 알 수 있다는 장점이 있습니다. 물론 UInt 타입으로 했을 때의 단점도 있습니다만 이 부분은 나중에 사용해보시면서 겪어보시는 걸 추천 드립니다 ㅎㅎ
3. 전역 변수와 지역 변수 용어에 대해서 오해가 있을 것 같습니다. guard의 경우에는 guard 문 이후에 사용을 할 수 있는 것이지 모든 전역에서 사용할 수 있는 것은 아니므로 `전역변수를 사용하고자 할 때`라는 용어는 오해를 일으킬 수 있을 것 같습니다. 또한 if 의 경우에도 해당 블록 내부에서 사용할 때 사용하는 것이지 함수의 모든 지역에서 사용하는 변수가 아니므로, 지역변수만이라는 용어는 오해를 일으킬 수 있을 것 같아요.
사용하는 데에 있어서 기준이 있다는 것이 멋있네요 👏

### 의문점에 대한 답변
1. 넵 좋습니다. 저 또한 프로젝트 기획서를 읽었을 때 두 분과 동일하게 JuiceMaker 타입 내부에 FruiteStore 인스턴스를 가지는 것으로 생각했습니다. 상속에 대해서 조금 더 공부를 해보시면 헷갈렸던 원인을 찾을 수 있을 것 같네요! 
2. 넵 수정하신 부분확인하고 코멘트 드렸습니다!

Jeon-Minsu committed:
style: 변수명 및 주석 네이밍 정정

bar-d committed:
refactor: 잘못된 양을 체크하는 구문을 빼내서 함수로 생성

bar-d commented:
반복되는 코드를 함수로 빼주었고 코멘트에 있던 내용들은 글을 남기거나 모두 수행을 한 것 같습니다...!!  
네이밍이... 참... 갈수록 어렵네요....🥲  
혹시나 빠뜨린 부분이나 부족한 부분이 있다면 다시한번 커멘트 남겨주세요 제이크!! @jryoun1   
잘 부탁드리겠습니다🙇‍♂️🙇‍♂️

jryoun1 commented:
수꿍 바드 step1 정말 고생하셨습니다 👏 
네이밍은 저도 지금도 쉽지 않아요 😂 
지금부터 고민하고 명확한 네이밍을 짓는 습관을 가질수 있도록 노력하시는 것을 추천드립니다! 

수정 및 답글 달아주신 부분들에 답변 드렸습니다. 
나머지는 step2에서 진행해도 될 것 같습니다! step2도 화이팅입니다 💪🏻
앗 머지한다는게 클로즈를 해버렸네요 step1은 머지 하도록 하겠습니다!
